### ISSUE TITLE
[RSpec/Rails] Including Celluloid in a class mess up connection with database

### ISSUE THREAD

**Issue by BastienDuplessier** [celluloid/celluloid#718]
Hello.
I have a little problem with celluloid, rspec and rails and I would like some help.

The problem is simple. I have a simple class with Celluloid included which count users.
```ruby
class UserCountCelluloid
  include Celluloid

  def count
    User.count
  end
end
```
And then, I have a test verifying if this method count the records.
```ruby
require 'rails_helper'
require 'user_count_celluloid'

describe UserCountCelluloid do
  before(:example) { User.create(email: 'foo@bar.com') }
  it "should count" do
    expect(UserCountCelluloid.new.count).to be 1
  end
end
```

But when I run RSpec, I get this : 
```
  1) UserCountCelluloid should count
     Failure/Error: expect(UserCountCelluloid.new.count).to be 1
     
       expected #<Fixnum:3> => 1
            got #<Fixnum:1> => 0
     
       Compared using equal?, which compares object identity,
       but expected and actual are not the same object. Use
       `expect(actual).to eq(expected)` if you don't care about
       object identity in this example.
     # ./spec/user_count_celluloid_spec.rb:9:in `block (2 levels) in <top (required)>'
```

Here's my spec_helper file, (my rails_helper is rails's default one).
```ruby
require 'celluloid/test'
require "celluloid/essentials"

RSpec.configure do |config|
  config.expect_with :rspec do |expectations|
    expectations.include_chain_clauses_in_custom_matcher_descriptions = true
  end

  config.mock_with :rspec do |mocks|
    mocks.verify_partial_doubles = true
  end
  config.profile_examples = 10
  config.order = :random
  Kernel.srand config.seed

  config.before(:each) do
    Celluloid.boot
  end
  config.after(:each) do
    Celluloid.shutdown
  end
end
```
This test passes if I remove `include Celluloid`
I only have this problem when running RSpec, if I try within a rails console, there's no problem.
I'm using Ruby 2.3.0, Rails 4.1.15, Celluloid 0.17.3 and RSpec Rails 3.4.0. The test database is SQLite3 (1.3.11).
Am I missing some configuration ? Is it a known bug ? Is this normal ? Is there some workaround ?

**Comment by digitalextremist** [comment]
It seems that `rails` has special handlers behind the scenes to cause `User` and `UserCountCelluloid` to relate to each other some how, because there is no explicit relationship between them that I can see.

**Comment by BastienDuplessier** [comment]
I've investigated a bit more the problem. It seems that it's not a Rails or RSpec problem but more a SQLite issue.
Seems that the database connection send back empty results when trying to access data through another thread. Trying to write (e.g. `User.create(email: "toto@mail.com")`) it raises this `ActiveRecord::StatementInvalid: SQLite3::BusyException: database is locked: INSERT INTO "users" ("created_at", "email", "updated_at") VALUES (?, ?, ?)`

The workaround I've found actually is to change our test database from SQLite to Mysql.
I've thought also about adding some `unless Rails.env.test?` in my code, but I find that solution quite ugly.
I'd rather have a another way to fix that issue without switching to Mysql in test env or adding some test specific code in my app.
Do you have any idea that can help me ?

**Comment by digitalextremist** [comment]
@chuckremes might know, but I don't think either of us are conversant with `Rails` convention; I do know he uses `SQLite` with `Celluloid` and he might have a quick opinion. Otherwise I will close this since you've identified it's not a `Celluloid` bug. Good luck!

**Comment by chuckremes** [comment]
@BastienDuplessier I use sqlite3 and celluloid together. You might want to look at your ActiveRecord configurations. There is probably an option to make sqlite3 access single threaded. If that is the default then trying to access the db from another thread will throw a lock exception.

I use the Sequel gem which has DB#connect(:single_threaded => true) as an option. Look for something similar in ActiveRecord. 

I hope this helps.

**Comment by BastienDuplessier** [comment]
Thanks for the answers, I'll check this.

**Comment by rally25rs** [comment]
I've seen the same thing with Postgres. In my case I tracked it back to our usage of the `database_cleaner` gem. By default that gem seems to use a transaction around the test, so that it can just roll it back after the test. However that means that other DB connections (in the Celluloid worker thread) can't read the data from the uncommitted transaction.
I ended up having to change my DatabaseCleaner strategy.
